### ISSUE TITLE
Add support for MultipartFileRecreatable from dio_smart_retry package

### ISSUE THREAD

**Issue by DmitryGaimaldinov** [trevorwang/retrofit.dart#698]
**Is your feature request related to a problem? Please describe.**
I use retrofit with [dio_smart_retry](https://pub.dev/packages/dio_smart_retry) package. I want to have ability to retry multipart requests with files. In dio_smart_retry this is done by MultipartFileRecreatable package. I wish retrofit had support to this. 

**Describe the solution you'd like**
I understand that retrofit package will never be dependent of another package like dio_smart_retry. But maybe there's a possibility to have another package which just dio_smart_retry for retrofit.

I've tried to fork retrofit locally and found out that make generator to generate MultipartFileRecreatable instead of MultipartFile is quite simple

```
final isRecreatable = r.peek('recreatable')?.boolValue ?? false; // I've just added `bool recreatable` property to Part annotation but of course this should be done another way

  final uploadFileInfo = refer(
          '${isRecreatable ? MultipartFileRecreatable : MultipartFile}.fromFileSync') // this is where code generated MultipartFile.fromFileSync
      .call([
    refer(p.displayName).property('path')
  ], {
    'filename': fileName,
    if (contentType != null)
      'contentType':
          refer('MediaType', 'package:http_parser/http_parser.dart')
              .property('parse')
              .call([literal(contentType)])
  });
```

But I faced a `FileSystemException: File closed` while retrying with `MultipartFileRecreatable` which happends because now generator generates `MultipartFile` from `File`. So I think it would be better to generate it not from file directrly but from some abstraction like `class FilePart { String filePath; String fileName; }`.

To sum up:
1) My idea is to create a package which adds dio_smart_retry support to retrofit.
2) Using `FilePart` (or another name) as `File` abstraction is safer that using `File` directly. But right now MultipartFile has no retry support so no bugs happen.

**Describe alternatives you've considered**
The only alternative right now is to use Dio instead of retrofit in cases when multipart requests with files needs to be retried.

**Additional context**
There were only one issue in dio_smart_retry about this case https://github.com/rodion-m/dio_smart_retry/issues/39. I guess there's no way to do it without retrofit support

**Comment by DmitryGaimaldinov** [comment]
I'm sorry. In first version of my comment above I mentioned that `FileSystemException: File closed` happens with my attempt to generate MultipartFileRecreatable instead of MultipartFile. But now I guess this exception happens only in my project, so I removed the mention of this exception
